lib/Conversion/TorchToLinalg/DataMovement.cpp

```diff
@@ -171,439 +172,348 @@ class ConvertAtenFlattenUsingIntsOp
-      return failure();
 
-    auto isValidReduction = [](int64_t expectedReductionProduct,
-                               ArrayRef<int64_t> arrayToReduce) -> bool {
-      if (llvm::count(arrayToReduce, kUnknownSize) > 0 ||
-          expectedReductionProduct == kUnknownSize)
-        return true;
-      return productReduce(arrayToReduce) == expectedReductionProduct;
-    };
+  // Try to fill in unknown dimensions in `to` using the statically known
+  // dimensions in `from`.
+  static void tryFillOneDirection(MutableArrayRef<int64_t> from,
+                                  MutableArrayRef<int64_t> to) {
 
-    if (xDims.size() == 1) {
-      if (!isValidReduction(xDims[0], yDims))
-        return failure();
-      xIndices.assign({0});
-      yIndices.assign(llvm::to_vector(llvm::seq<int64_t>(0, yDims.size())));
-      return success();
-    } else if (yDims.size() == 1) {
-      if (!isValidReduction(yDims[0], xDims))
-        return failure();
-      yIndices.assign({0});
-      xIndices.assign(llvm::to_vector(llvm::seq<int64_t>(0, xDims.size())));
-      return success();
+    auto fromProduct = productReduce(from);
+
+    SmallVector<uint64_t> unknowns;
+    for (auto [i, dim] : llvm::enumerate(to)) {
+      if (dim == kUnknownSize)
+        unknowns.push_back(i);
     }
-    return failure();
-  }
 
-  // Starting from the beginning of the dims arrays, this helper finds the
-  // smallest set of consecutive dims in each array such that the product of the
-  // dim sizes in the two subsets is equal. The indices arrays are populated
-  // with the indices of the dims arrays that correspond to the subsets found.
-  //
-  // An error is returned if two subsets of dims with total number of elements
-  // equal to each other is not found.
-  static LogicalResult mapStaticallyKnownDims(ArrayRef<int64_t> xDims,
-                                              ArrayRef<int64_t> yDims,
-                                              SmallVector<int64_t> &xIndices,
-                                              SmallVector<int64_t> &yIndices) {
-    if (xDims.empty() || yDims.empty())
-      return failure();
-    int64_t xTotalSize = xDims[0];
-    int64_t yTotalSize = yDims[0];
-    SmallVector<int64_t> xIndicesResult({0});
-    SmallVector<int64_t> yIndicesResult({0});
-    size_t nextXIndex = 1;
-    size_t nextYIndex = 1;
-    while (xTotalSize != yTotalSize) {
-      if (xTotalSize < yTotalSize) {
-        if (nextXIndex == xDims.size() || xDims[nextXIndex] == kUnknownSize)
-          return failure();
-        xTotalSize *= xDims[nextXIndex];
-        xIndicesResult.push_back(nextXIndex++);
-      } else {
-        if (nextYIndex == yDims.size() || yDims[nextYIndex] == kUnknownSize)
-          return failure();
-        yTotalSize *= yDims[nextYIndex];
-        yIndicesResult.push_back(nextYIndex++);
+    auto nUnknowns = unknowns.size();
+
+    auto toProduct =
+        productReduce(to) / std::pow<int64_t>(kUnknownSize, nUnknowns);
```

Comment:
I'm not sure about this change. What if torch-mlir decides to change `kUnknownSize` to, say, INT64_MIN? This is what happened in upstream llvm and this could be a tricky bug to catch.

Reply:
+1 to that then. My understanding of those kinds of sentinel constants is that you shouldn't assume they are a specific value, just that they are static, meaning arithmetic with them is probably a bad idea. If this was present in the old code then removing the division would definitely be nice cleanup!